### ISSUE TITLE
Manage to get context manager of context to manage context automatically

### ISSUE THREAD

**Issue by McSinyx** [McSinyx/palace#52]
The conventional purpose of context managers is to maintain a context within the `with` block.  Thus, the current interface is rather redundant:
```python
with (Device() as dev, Context(dev) as ctx, Buffer(ctx, 'resource') as buf,
      Source(ctx) as src, SourceGroup(ctx) as grp,
      Effect(ctx) as fx, AuxiliaryEffectSlot(ctx) as aux):
    dec = Decoder(ctx, 'resource')
```
since all these instantiations requires `ctx` to be current to run anyway.

Alternatively, we can allow the initializers to use the current context, i.e.
```python
with (Device() as dev, Context(dev) as ctx, Buffer('resource') as buf,
      Source() as src, SourceGroup() as grp,
      Effect() as fx, AuxiliaryEffectSlot() as aux):
    dec = Decoder('resource')
```

Effectively, this means the following should also runs without any error:
```python
dev = Device()
ctx = Context(dev)
use_context(ctx)
buf = Buffer('resource')
src = Source()
grp = SourceGroup()
fx = Effect()
aux = AuxiliaryEffectSlot()
dec = Decoder('resource')
# closure code
```

If we decide to do this, `context` should be moved to an optional argument which is `None` by default.  Additionally, it's also time to decide what to do with real attributes like `listener` and `device` after the destruction of these objects (should we set them to `None`?).

Edit: `Decoder.smart` is subject to change.

**Comment by McSinyx** [comment]
Note that with this implemented, we can (again) let decoders create there own sources for playback (yay!).

**Comment by McSinyx** [comment]
Related to this, the `thread` parameter in `current_context`, `use_context` should be able to fallback to a module-state controlled by a context manager, i.e.
```python
with thread(False):
    current_context()             # context current to all threads
    current_context(thread=False) # context current to all threads
    current_context(thread=True)  # context current to calling thread

with thread(True):
    current_context()             # context current to calling thread
    current_context(thread=False) # context current to all threads
    current_context(thread=True)  # context current to calling thread
```

Edit: done in 2a3bda152fe266ebbb327d3f15d96e038463

**Comment by McSinyx** [comment]
Should be solved by 49072f101e159c0e27b53b98f9367cecf57acf0f.
